Fix refresh problem exposed by issue 30315:

tbitonti opened the pull request:
The fix for #30315 exposed a problem in annotation caching, where the overall results were not always updated.

The fix for 30315 has been delivered, but was disabled because of the additional problem.

This update fixes the problem which was discovered, and re-enables the 30315 fix code.

Fixes #30315

- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

################################################################################################

### Technical background

During application startup, when the application does not have an application deployment descriptor (application.xml), there is a step of discovering modules.  This requires scanning (non-library) application jar files for annotations, since one of the ways that jar files are determined to be application client jars or EJB jars (instead of being simple utility jars) is by detecting application client and EJB annotations.

The step of scanning caused a partial update module annotations data: The data for individual jar files was refreshed, however, the overall results for each module were not marked as invalid, and were not refreshed.  The failure to update the overall results for modules occurred because the initial annotation scan did not type the overall results as being for a module.

Later in application startup, scan results for the jar files is used a second time.  The second use typed the scan results as being for a module.  This creates (or loads) a different layer of information -- module information -- for the jar files.  If the module information is reloaded, the information for the JAR was refreshed, however the overall module information was not flagged as being invalid.

The fix was to record container details -- the container last update time and size -- to the overall results information.  That enables a check to tell if the overall results are invalid if any container was changed, which triggers a necessary reload.

tbitonti committed:
Fix refresh problem exposed by issue 30315:
Have the containers list record container time and size and check this when looking for changes.
Update the containers list persistence format to store the time and size.
Re-enable the 30315 fix.

tbitonti commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=dcadbd1e-47a0-4b65-90a1-f660070bdb52) - :white_check_mark: completed successfully!)

_Note: Target locations of links might be accessible only to IBM employees._

tbitonti commented:
Need to update the "sameAs" code to properly handle "unavailable" and "unrecorded" and have these appear to be different.
Will add a commit with this update.

tbitonti committed:
Correctly handle uncomparable container signatures.

tbitonti committed:
Change incomplete 'PHXXXXX' comments to 'Issue 30315'.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/targets/internal/TargetsScannerOverallImpl.java:
```diff
@@ -1747,6 +1754,11 @@ protected void validInternalResults() {
+                //     if ( cachedTable != null ) {
+                //         System.out.println("Read table [ " + cachedTable.getName() + " ] Stamp [ " + cachedTable.getStamp() + " ]");
+                //     }
+                // }
```
```suggestion
```
Any reason to leave in commented out debug code?

tbitonti replied:
Removed.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/targets/internal/TargetsScannerBaseImpl.java:
```diff
@@ -73,6 +73,8 @@ public TargetsScannerBaseImpl(
         
         this.hashText = getClass().getSimpleName() + "@" + Integer.toHexString(hashCode());
 
+        // (new Throwable(this.hashText)).printStackTrace(System.out);
+
```
```suggestion
```
Remove of debug code?

tbitonti replied:
Removed.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/targets/cache/internal/TargetCacheImpl_WriterBinary.java:
```diff
@@ -408,16 +408,17 @@ public void writeFragment(TargetsTableContainersImpl containerTable) throws IOEx
+            ScanPolicy policy = containerTable.getPolicy(name);
+            
+            writeCompact(NAME_BYTE, useName);
+            writeCompact(SIGNATURE_BYTE, signature);
```
You putting it before policy instead of after policy in order to know if it is old or new form?  What happens if it is written in the new format and then read by downlevel code.  Shouldn't it be after policy for that reason?

tbitonti replied:
The format is determined by the table header, which was read previously.  An attempt to read the table in the version 2.0 format by the old 1.0 reader will fail.  See 'com.ibm.ws.annocache.targets.cache.internal.TargetCacheImpl_ReaderBinary.requireHeader(String, String[], int[])',
around line 253.  Older code has a similar 'requireHeader' which accepts only 1.0 as the table version.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/targets/TargetsTableContainers.java:
```diff
@@ -65,7 +76,19 @@ public interface TargetsTableContainers {
      * @param name The name to add.
      * @param policy The scan policy associated with the name.
      */
+    @Deprecated    
```
Should we just remove the old methods instead of deprecating them?  They are internal methods / classes that cannot be used outside of internal liberty code.

tbitonti replied:
Removing.

tbitonti committed:
Remove incorrect tests from container table.

tbitonti commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=20a64331-b9a6-41ab-b83a-11b3dd48602c) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/targets/internal/TargetsTableContainersImpl.java:
```diff
@@ -165,10 +220,24 @@ public void log(Logger useLogger) {
     //
 
     public boolean sameAs(TargetsTableContainersImpl otherTable) {
+        return ( sameAs(otherTable, !DO_DETAIL ) == null );
```
`!DO_DETAIL` seems wrong for describe.  Should it be the opposite?

tbitonti replied:
This method is being updated.  A new review will be needed after the updates are posted.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/targets/cache/internal/TargetCacheImpl_Reader.java:
```diff
@@ -228,13 +311,15 @@ public List<TargetCache_ParseError> parse(TargetsReader useReader)
         if ( logger.isLoggable(Level.FINER) ) {
             logger.logp(Level.FINER, CLASS_NAME, methodName, "Expected table [ " + expectedTableTag + " ]");
-            logger.logp(Level.FINER, CLASS_NAME, methodName, "Expected table version [ " + expectedTableVersion + " ]");            
+            logger.logp(Level.FINER, CLASS_NAME, methodName, "Accepted table version [ " + acceptedTableVersions + " ]");            
```
`acceptedTableVersions` won't give the toString you are wanting.  Maybe use Arrays.toString() around it?

tbitonti replied:
Fixed.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/targets/cache/internal/TargetCacheImpl_Reader.java:
```diff
@@ -335,6 +420,19 @@ public TargetsReader() {
+         * @return The versions which are accepted on the version line.
+         */
+        public String[] getTableVersions() {
+            return new String[] { getTableVersion() };
```
Should this method be abstract like getTableVersion?

tbitonti replied:
Most tables have a single version.  This provides a default implementation for those table types.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/targets/cache/internal/TargetCacheImpl_Reader.java:
```diff
@@ -383,16 +481,20 @@ public void handleTrailer() {
 
     public class ContainerTableReader extends TargetsReader {
-        public ContainerTableReader(TargetsTableContainers containerTable) {
+        public ContainerTableReader(TargetsTableContainers containerTable, TargetCacheImpl_Reader reader) {
```
since it is an inner class it has a pointer to its outer class.  If it doesn't use it, should the class be made static now that the Reader is passed into the constructor?

tbitonti replied:
Not sure why I added that.  The reader parameter is not necessary.  Resolving by removing the parameter.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/targets/cache/internal/TargetCacheImpl_ReaderBinary.java:
```diff
@@ -218,6 +219,11 @@ public String requireCompact() throws IOException {
     //
 
     public void requireHeader(String tableName, String tableVersion) throws IOException {
```
Is this method needed any longer?

tbitonti replied:
'requireCompact()', with no parameters, is used by 'requireCompact(byte)' and by 'readClassTargets'.
'requireCompact(byte)' expects:
`byte large-int`
'readClassTargets' expects:
`byte small-int large-int*`
In both cases, the read of a large-int is the read of an offset into a string table.  'requireCompact' validates the read offset and does a lookup of the value.

jhanders34 replied:
I was asking about the `requireHeader` method, not the `requireCompact()` method.

jhanders34 replied:
You cleaned up the method so I marked it resolved.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/targets/cache/internal/TargetCacheImpl_ReaderBinary.java:
```diff
@@ -241,12 +247,20 @@ public void requireHeader(String tableName, String tableVersion) throws IOExcept
             throw new IOException(
                 "Unexpected table version [ " + actualVersion + " ]:" +
-                " Expecting table version [ " + tableVersion + " ]" +
+                " Expecting table versions [ " + tableVersions + " ]" +
```
Again toString of `tableVersions` will not be very useful.  Surround with Array.toString()?

tbitonti replied:
Fixed.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/targets/cache/internal/TargetCacheImpl_ReaderBinary.java:
```diff
@@ -466,7 +480,8 @@ public void readEntire(TargetsTableContainersImpl containerTable) throws IOExcep
     public void readFragment(TargetsTableContainersImpl containerTable) throws IOException {
-        requireContainersHeader();
+        String readVersion = requireContainersHeader();
+        boolean readStamps = readVersion.equals(TargetCache_BinaryConstants.CONTAINER_TABLE_VERSION_20);
```
In the non Binary class you did greater than or equal to.  Should that be done here as well?

tbitonti replied:
Fixing.

tbitonti committed:
Additional adjustments.

tbitonti committed:
Back out 'earliestStamp' changes.
Removing compiler warnings.

tbitonti committed:
Review adjustments:
Remove debugging code.
Remove deprecated methods.
Correct requireHeader API.
Correct 'requireHeader' version check to use >=.
Clarify TargetsTableContainerImpl.sameAs.

tbitonti committed:
Update copyrights.

tbitonti commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=0eb56e45-d29e-4fda-ab27-624d678c86ee) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/targets/cache/internal/TargetCacheImpl_Reader.java:
```diff
@@ -173,41 +211,87 @@ protected String parseLine() {
+            }
+        }
+        if ( !matched ) {
+            addParseError("Value [ " + parsedValue + " ] does not match [ " + allowedValues + " ] for class [ " + getClass().getName() + " ]");
```
```suggestion
            addParseError("Value [ " + parsedValue + " ] does not match [ " + Arrays.toString(allowedValues) + " ] for class [ " + getClass().getName() + " ]");
```

tbitonti replied:
Agh, missed one.

jhanders34 reviewed (changes requested):
I have some more looking to do for the new code, but I figured I would give you the one change for sure.

tbitonti committed:
Update dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/targets/cache/internal/TargetCacheImpl_Reader.java

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/wsspi/anno/info/MethodInfo.java:
```diff
@@ -133,12 +133,12 @@ public interface MethodInfo extends Info {
      * @return The in-order parameter annotation collections of parameters
      *         of this method.
      */
-    public List<List<? extends AnnotationInfo>> getParameterAnnotations();
```
This build failed because this is an SPI so you can't just delete a method and add a new one without updating version.  I think you can leave it the way it was in order to not require a SPI version update.

jhanders34 replied:
I see you added it back and also added all the other `@Deprecated` methods in the latest commits.  Only this one method needed to be added back since it is the only one that is on a SPI interface.

tbitonti replied:
Fixing MethodInfo (in anno) to be as it was before.  The update is not needed here, just in annocache.
(I missed that the prior comment was re: the copy under "anno".)

tbitonti committed:
Remove deprecated APIs from unit tests.
Remove unneeded EE API imports from unit tests.
Cleanup unit test warnings.

tbitonti commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=4626c1f5-94b4-47f8-ad25-0a41d17396d1) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

jhanders34 commented on dev/com.ibm.ws.anno/test/com/ibm/ws/anno/test/cases/SecFVT_Servlet30_Inherit_war_Test.java:
```diff
@@ -92,7 +90,7 @@ public void testSecFVT_Servlet30_Common_war_detail_direct() throws Exception {
     }
 
-    public static final String SERVLET_SECURITY_CLASS_NAME = ServletSecurity.class.getName();
+    public static final String SERVLET_SECURITY_CLASS_NAME = "com.ibm.ws.anno.test.cases.ServletSecurity";
```
```suggestion
    public static final String SERVLET_SECURITY_CLASS_NAME = "javax.servlet.annotation.ServletSecurity";
```
Shouldn't it be that?

jhanders34 commented on dev/com.ibm.ws.anno/test/com/ibm/ws/annocache/test/info/InfoStore_Test.java:
```diff
@@ -420,10 +419,6 @@ public void testParameterAndAnnoOrder() {
         Assert.assertEquals(4, parmTypes.size());
         Assert.assertEquals(Arrays.asList("int", String.class.getName(), List.class.getName(), "long"), parmTypes);
 
-        @SuppressWarnings("deprecation")
-        List<List<? extends com.ibm.wsspi.anno.info.AnnotationInfo>> parameterAnnos = mi.getParameterAnnotations();
-        Assert.assertEquals(4, parameterAnnos.size());
-
```
Probably going to have to put this back.

tbitonti replied:
Changed back.

tbitonti committed:
Update dev/com.ibm.ws.anno/test/com/ibm/ws/anno/test/cases/SecFVT_Servlet30_Inherit_war_Test.java

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

tbitonti committed:
Add back deprecated APIS (avoids APICHK errors).

tbitonti committed:
Prevent compiler deprecation warning.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/wsspi/anno/info/MethodInfo.java:
```diff
@@ -137,8 +137,18 @@ public interface MethodInfo extends Info {
+     * @return The in-order parameter annotation collections of parameters
+     *         of this method.
+     */
+    public List<? extends List<? extends AnnotationInfo>> getParmAnnotations();
```
I have no idea why you are adding this method in and deprecating the one above it.  They are the same thing.  If you are going to add this method to the SPI class, then you need to update the minor version.  I just think this getParmAnnotations() stuff just should go away.

tbitonti replied:
List\<? extends T\> is different than List\<T\>.
The use of a wildcard provides a bound on the generic type without exactly specifying the type.  When used with a collection type, the effect is for the collection to be non-modifiable, because parameters are contravariant.
For this method, I think the new method was added because of how the annocache types extend the anno types.  The original getParameters could not be extended in the annocache copy of the type.

jhanders34 replied:
Well you changed from `List<List>` to `List<? extends List>`. So there is a different `List` type that you are using I assume then.  The annocache classes are not exposed as SPI, so not sure if you need to have this new method added in the non annocache version of the interface.

tbitonti committed:
Remove 'getParameterAnnotations" from "anno" MethodInfo.

tbitonti commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=954e906d-8bce-4574-89d5-ec9ad3df8991) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tbitonti commented:
Problem found:

```
<failure message="2025-05-16-07:06:59:004 [Unexpected FFDC reporting java.io.IOException was found (count = 1): /home/ci/Build/workspace/ebcTestRunner/dev/autoFVT/image/output/wlp/usr/servers/com.ibm.ws.app.manager.wab.installer/logs/ffdc/ffdc_25.05.16_07.06.58.0.log 
>------Start of DE processing------ = [5/16/25, 7:06:58:397 EDTâ¯] 
>Exception = java.io.IOException >Source = com.ibm.ws.annocache.targets.cache.internal.TargetCacheImpl_DataBase 
>probeid = 815 >Stack Dump = 
java.io.IOException: Read [ 4 ] at [ 0 ] past end of file [ 0 ] on [ /home/ci/Build/workspace/ebcTestRunner/dev/autoFVT/image/output/wlp/usr/servers/com.ibm.ws.app.manager.wab.installer/workarea/org.eclipse.osgi/183/data/anno/A_test.wab1_A/C_test.wab1_C.stamp ]
> at com.ibm.ws.annocache.util.internal.UtilImpl_ReadBufferPartial.read(UtilImpl_ReadBufferPartial.java:246) 
> at com.ibm.ws.annocache.targets.cache.internal.TargetCacheImpl_ReaderBinary.requireMagic(TargetCacheImpl_ReaderBinary.java:131) 
> at com.ibm.ws.annocache.targets.cache.internal.TargetCacheImpl_ReaderBinary.requireBegin(TargetCacheImpl_ReaderBinary.java:142) 
> at com.ibm.ws.annocache.targets.cache.internal.TargetCacheImpl_ReaderBinary.<init>(TargetCacheImpl_ReaderBinary.java:76) 
> at com.ibm.ws.annocache.targets.cache.internal.TargetCacheImpl_Factory.createBinaryReader(TargetCacheImpl_Factory.java:330) 
> at com.ibm.ws.annocache.targets.cache.internal.TargetCacheImpl_DataBase.createBinaryReader(TargetCacheImpl_DataBase.java:777) 
> at com.ibm.ws.annocache.targets.cache.internal.TargetCacheImpl_DataBase.readBinary(TargetCacheImpl_DataBase.java:805) 
> at com.ibm.ws.annocache.targets.cache.internal.TargetCacheImpl_DataCon.readStamp(TargetCacheImpl_DataCon.java:305) 
> at com.ibm.ws.annocache.targets.cache.internal.TargetCacheImpl_DataCon.isValid(TargetCacheImpl_DataCon.java:639) 
> at com.ibm.ws.annocache.targets.internal.TargetsScannerOverallImpl.validInternalContainer(TargetsScannerOverallImpl.java:860) 
> at com.ibm.ws.annocache.targets.internal.TargetsScannerOverallImpl.validInternalContainers(TargetsScannerOverallImpl.java:1027) 
> at com.ibm.ws.annocache.targets.internal.TargetsScannerOverallImpl.validInternalContainers_Select(TargetsScannerOverallImpl.java:978) 
> at com.ibm.ws.annocache.targets.internal.TargetsScannerOverallImpl.validInternalUnresolved(TargetsScannerOverallImpl.java:1570) 
> at com.ibm.ws.annocache.targets.internal.TargetsScannerOverallImpl.validInternal(TargetsScannerOverallImpl.java:2094) 
> at com.ibm.ws.annocache.targets.internal.AnnotationTargetsImpl_Targets.ensureInternalResults(AnnotationTargetsImpl_Targets.java:730) 
> at com.ibm.ws.annocache.targets.internal.AnnotationTargetsImpl_Targets.getSeedTable(AnnotationTargetsImpl_Targets.java:559) 
> at com.ibm.ws.annocache.targets.internal.AnnotationTargetsImpl_Targets.getAnnotatedClasses(AnnotationTargetsImpl_Targets.java:1117) 
> at com.ibm.ws.annocache.targets.internal.AnnotationTargetsImpl_Targets.getAnnotatedClasses(AnnotationTargetsImpl_Targets.java:1169) 
> at com.ibm.ws.container.service.annocache.internal.FragmentAnnotationsImpl.selectAnnotatedClasses(FragmentAnnotationsImpl.java:52) 
> at com.ibm.ws.webcontainer.osgi.container.config.WebAppConfiguratorHelper.configureFromAnnotations(WebAppConfiguratorHelper.java:902) 
> at com.ibm.ws.webcontainer.osgi.container.config.WebAppConfigurator.configure(WebAppConfigurator.java:463) 
> at com.ibm.ws.webcontainer.osgi.container.config.WebAppConfigurationAdapter.adapt(WebAppConfigurationAdapter.java:121) 
> at com.ibm.ws.webcontainer.osgi.container.config.WebAppConfigurationAdapter.adapt(WebAppConfigurationAdapter.java:45) 
> at com.ibm.ws.adaptable.module.internal.AdapterFactoryServiceImpl.adapt(AdapterFactoryServiceImpl.java:202) 
> at com.ibm.ws.adaptable.module.internal.AdaptableContainerImpl.adapt(AdaptableContainerImpl.java:176) 
> at com.ibm.ws.webcontainer.osgi.WebContainer.createModuleMetaData(WebContainer.java:872) 
> at com.ibm.ws.app.manager.module.internal.ModuleHandlerBase.createModuleMetaData(ModuleHandlerBase.java:65) 
> at com.ibm.ws.app.manager.module.internal.SimpleDeployedAppInfoBase$ModuleContainerInfoBase.createModuleMetaData(SimpleDeployedAppInfoBase.java:222) 
> at com.ibm.ws.app.manager.module.internal.SimpleDeployedAppInfoBase.preDeployApp(SimpleDeployedAppInfoBase.java:608) 
> at com.ibm.ws.app.manager.module.internal.SimpleDeployedAppInfoBase.installApp(SimpleDeployedAppInfoBase.java:584) 
> at com.ibm.ws.app.manager.wab.internal.WABInstaller$WABDeployedAppInfo.installApp(WABInstaller.java:1570) 
> at com.ibm.ws.app.manager.wab.internal.WABInstaller.installIntoWebContainer(WABInstaller.java:522) 
...
```

tbitonti commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=3622d9a4-0e8e-4172-875f-46739c7d38ca) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tbitonti commented:
This:Read [ 4 ] at [ 0 ] past end of file [ 0 ] on [ /home/ci/Build/workspace/ebcTestRunner/dev/autoFVT/image/output/wlp/usr/servers/com.ibm.ws.app.manager.wab.installer/workarea/org.eclipse.osgi/183/data/anno/A_test.wab1_A/C_test.wab1_C.stamp ] > at com.ibm.ws.annocache.util.internal.UtilImpl_ReadBufferPartial.read(UtilImpl_ReadBufferPartial.java:246) > at
Shows that the file "C_test.wab1_C.stamp" is empty at the time of read.
That would be caused by a prior write failure, or by a read concurrent with the creation of the file.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/targets/TargetsTableContainers.java:
```diff
@@ -67,6 +67,17 @@ public interface TargetsTableContainers {
     ScanPolicy getPolicy(String name);
 
     //
+    
+    /**
+     * <p>Add a name and a scan policy to this container table.</p>
+     *
+     * <p>Names are kept in the order in which they were added.</p>
+     *
+     * @param name The name to add.
+     * @param policy The scan policy associated with the name.
+     */
+    @Deprecated    
+    void addName(String name, ScanPolicy policy);    
```
```suggestion
```
No need to add it back.

tbitonti replied:
Removed.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/targets/TargetsTableContainers.java:
```diff
@@ -79,6 +90,19 @@ public interface TargetsTableContainers {
+     */
+    @Deprecated
+    void addNameAfter(String name, ScanPolicy policy, String afterName);
+   
```
```suggestion
```
No need to add it back.

tbitonti replied:
Removed.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/targets/TargetsTableContainers.java:
```diff
@@ -92,6 +116,19 @@ public interface TargetsTableContainers {
+     */
+    @Deprecated
+    void addNameBefore(String name, ScanPolicy policy, String beforeName);
+    
```
```suggestion
```
No need to add it back.

tbitonti replied:
Removed.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/targets/internal/TargetsTableContainersImpl.java:
```diff
@@ -91,13 +91,32 @@ private void putSignature(String name, String signature) {
+        names.add(name);
+        policies.put(name, policy);
+    }
+    
```
```suggestion
```
No need to add it back.

tbitonti replied:
Removed.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/targets/internal/TargetsTableContainersImpl.java:
```diff
@@ -91,13 +91,32 @@ private void putSignature(String name, String signature) {
+        names.add(addOffset + 1, name);
+        policies.put(name, policy);
+    }
+    
```
```suggestion
```
No need to add it back.

tbitonti replied:
Removed.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/targets/internal/TargetsTableContainersImpl.java:
```diff
@@ -110,6 +129,18 @@ public void addNameAfter(String name, String signature, ScanPolicy policy, Strin
+        names.add(addOffset, name);
+        policies.put(name, policy);
+    }
+
```
```suggestion
```
No need to add it back

tbitonti replied:
Removed.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/util/internal/UtilImpl_BidirectionalMap.java:
```diff
@@ -78,6 +78,15 @@ public Util_Factory getFactory() {
+    }
+
+    //
+
```
```suggestion
```
No need to add it back

tbitonti replied:
Removed.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/ws/annocache/util/internal/UtilImpl_EmptyBidirectionalMap.java:
```diff
@@ -68,6 +68,15 @@ public UtilImpl_Factory getFactory() {
+    }
+
+    //
+
```
```suggestion
```
No need to add it back

tbitonti replied:
Removed.

jhanders34 commented on dev/com.ibm.ws.anno/src/com/ibm/wsspi/annocache/util/Util_BidirectionalMap.java:
```diff
@@ -31,6 +31,17 @@ public interface Util_BidirectionalMap extends com.ibm.wsspi.anno.util.Util_Bidi
+    boolean getEnabled();
+
+    //
+
```
```suggestion
```
No need to add it back.

tbitonti replied:
Removed.

tbitonti committed:
Use ApplicationInfo.getName() for annotations cache data instead of AnnotationInfo.getDeployedName().

tbitonti committed:
BiDi map 'getIsEnabled' fixes.  Deprecation fixes.

tbitonti committed:
Merge branch 'integration' into 30315-refresh-fix

tbitonti commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=118385d7-eb1d-4383-9c6d-44fe62400e65) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tbitonti commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=8b7ba85f-9891-48ea-8dad-8fc51e30e83c) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tbitonti commented:
#build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=07da6924-a609-477b-afe9-6d3df75da6ea) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

tbitonti commented:
Failures seem unrelated.
